### PR TITLE
feat: Parallelise picture processing

### PR DESCRIPTION
**Pull request by mfornet**
Use rayon to parallelise picture processing. Speed up 6.5x (from 31.88s to 4.93s 95 images of size ~10Mb each in a local test).

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -223,77 +225,87 @@ fn main() -> Result<()> {
...
-        });
-    }
+            // Store
+            Image {
+                filename_full,
+                filename_thumb,
+            }
+        })
+        .collect::<Vec<_>>();
+    // for f in &image_files {}
```
Comment (mfornet) [inline src/main.rs]:
```suggestion

```

File: src/main.rs [hunk 2]
```diff
@@ -223,77 +225,86 @@ fn main() -> Result<()> {
...
-            let options = zip::write::FileOptions::default()
-                .compression_method(zip::CompressionMethod::Stored);
-            if let Some(ref mut zipwriter) = zipfile {
-                zipwriter.start_file(&filename_full, options)?;
-                zipwriter.write_all(&fs::read(&full_path)?)?;
+                // Add file to ZIP
+                let options = zip::write::FileOptions::default()
+                    .compression_method(zip::CompressionMethod::Stored);
+
+                zipfile.lock().unwrap().iter_mut().for_each(|zipwriter| {
```
Comment (dbrgn) [inline src/main.rs]: I'm a bit confused why `iter_mut().for_each(...)` is used here. After all, this is an `Arc<Mutex<ZipWriter>>`, right?

**Review by dbrgn — COMMENTED** [review]
Nice, thanks for the useful contribution!

image

It's much faster, but the error handling is a bit of a step back:

- Many anyhow errors that were previously propagated are now simply `unwrap()`s. In general panics would be fine for a CLI, if the calls are replaced with `except(...)`s with a better error message, but:
- If processing a single image fails (e.g. due to a permission error, can be tested by making one of the output files only writable by root), then processing of other images does not abort. However, once all images are processed, the program simply aborts without an error message and without writing the index file.

I think ideally every Rayon task would return a result, and if processing of even one image fails, then the entire parallel pipeline should be aborted with an error message.

I don't know what Rayon has to offer with regards to error handling. Do you think these things could be added?

File: src/main.rs (hunk 2 above)
Comment (mfornet) [reply src/main.rs]: Fixed!

**Comment by mfornet** [comment]
I'm propagating errors in the parallel iterator now and already checked that the iterator won't necessarily finish if there is an earlier error.

**Comment by atteism** [comment]
Thank you Danilo, just found this and am trying it for the first time. I have processed half of the 650 photos (45 MP) in 25 minutes, so mfornet's patch with multithread support would be really good to get merged :-)

**Comment by dbrgn** [comment]
@mfornet great, thanks for the update! Sorry, I missed it previously.

Error handling is much better now! There are error messages that can be improved, but I can do that after merging.

**Comment by dbrgn** [comment]
Improved error handling implemented in #8.

**Comment by dbrgn** [comment]
@atteism @mfornet this change is now released in v1.2.0!

**Pull request merged**